core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java

```diff
@@ -17,87 +17,150 @@
       return new NoNodeAvailableException();
     } else {
-      return new AllNodesFailedException(ImmutableMap.copyOf(errors));
+      return new AllNodesFailedException(groupByNode(errors.entrySet()));
     }
   }
 
   @NonNull
-  public static AllNodesFailedException fromErrors(
-      @Nullable List<Map.Entry<Node, Throwable>> errors) {
-    Map<Node, Throwable> map;
+  public static AllNodesFailedException fromErrors(@Nullable List<Entry<Node, Throwable>> errors) {
     if (errors == null || errors.isEmpty()) {
-      map = null;
+      return new NoNodeAvailableException();
     } else {
-      ImmutableMap.Builder<Node, Throwable> builder = ImmutableMap.builder();
-      for (Map.Entry<Node, Throwable> entry : errors) {
-        builder.put(entry);
-      }
-      map = builder.build();
+      return new AllNodesFailedException(groupByNode(errors));
     }
-    return fromErrors(map);
   }
 
-  private final Map<Node, Throwable> errors;
+  private final Map<Node, List<Throwable>> errors;
 
+  /** @deprecated Use {@link #AllNodesFailedException(String, ExecutionInfo, Iterable)} instead. */
+  @Deprecated
   protected AllNodesFailedException(
       @NonNull String message,
       @Nullable ExecutionInfo executionInfo,
       @NonNull Map<Node, Throwable> errors) {
+    this(message, executionInfo, groupByNode(errors.entrySet()));
+  }
+
+  protected AllNodesFailedException(
+      @NonNull String message,
+      @Nullable ExecutionInfo executionInfo,
+      @NonNull Iterable<Entry<Node, List<Throwable>>> errors) {
     super(message, executionInfo, null, true);
-    this.errors = errors;
+    this.errors = toDeepImmutableMap(errors);
   }
 
-  private AllNodesFailedException(Map<Node, Throwable> errors) {
+  private AllNodesFailedException(Collection<Entry<Node, List<Throwable>>> errors) {
     this(
         buildMessage(
             String.format("All %d node(s) tried for the query failed", errors.size()), errors),
         null,
         errors);
   }
 
-  private static String buildMessage(String baseMessage, Map<Node, Throwable> errors) {
+  private static String buildMessage(
+      String baseMessage, Collection<Entry<Node, List<Throwable>>> errors) {
     int limit = Math.min(errors.size(), 3);
-    String details =
-        Joiner.on(", ").withKeyValueSeparator(": ").join(Iterables.limit(errors.entrySet(), limit));
-
+    Iterator<Entry<Node, List<Throwable>>> iterator = errors.iterator();
+    StringBuilder details = new StringBuilder();
+    while (iterator.hasNext()) {
+      Entry<Node, List<Throwable>> entry = iterator.next();
+      details.append(entry.getKey()).append(": ").append(entry.getValue());
+      if (iterator.hasNext()) {
+        details.append(", ");
+      }
+    }
     return String.format(
-        baseMessage + " (showing first %d, use getErrors() for more: %s)", limit, details);
+        baseMessage + " (showing first %d, use getAllErrors() for more): %s", limit, details);
```

Comment:
Nit: I allowed myself to move the closing parenthesis before the errors list, which can be very long; otherwise, it seems that the opening parenthesis is unbalanced.